### PR TITLE
Added support for multiple data-stores.

### PR DESCRIPTION
**Pull request by ipapapa**
Added changes for multiple data stores. Modified the .yml files. redis = 1, memcached = 2, any other can be > 2.

### CODE REVIEW

File: src/dyn_conf.h [hunk 1]
```diff
@@ -52,7 +52,7 @@
 #define CONF_DEFAULT_TIMEOUT                 500
 #define CONF_DEFAULT_LISTEN_BACKLOG          512
 #define CONF_DEFAULT_CLIENT_CONNECTIONS      0
-#define CONF_DEFAULT_REDIS                   false
+#define CONF_DEFAULT_DATASTORE				 0
```
Comment (shailesh33) [inline src/dyn_conf.h]: DATA_REDIS?

File: src/dyn_conf.h [hunk 2]
```diff
@@ -117,7 +117,7 @@ struct conf_pool {
     int                timeout;               /* timeout: */
     int                backlog;               /* backlog: */
     int                client_connections;    /* client_connections: */
-    int                redis;                 /* redis: */
+    int                data_store;            /* data_store: */
```
Comment (shailesh33) [inline src/dyn_conf.h]: enum please?

File: src/dyn_core.h [hunk 3]
```diff
@@ -69,6 +69,9 @@
 #define DN_EAGAIN   -2
 #define DN_ENOMEM   -3
 
+#define DATA_REDIS		 	0	/* Data store is Redis */
```
Comment (shailesh33) [inline src/dyn_core.h]: ya an enum makes complete sense here.

File: src/dyn_dnode_msg.c [hunk 4]
```diff
@@ -517,10 +525,18 @@ dyn_parse_req(struct msg *r)
 		} else if (r->dyn_state == DYN_POST_DONE) {
 			struct mbuf *last_buf = STAILQ_LAST(&r->mhdr, mbuf, next);
 			if (last_buf->read_flip == 1) {
-				if (r->redis)
+				if (r->data_store == DATA_REDIS) {
```
Comment (shailesh33) [inline src/dyn_dnode_msg.c]: There is a lot of code repeatition. You could convert it to a wrapper function and use that here.

File: src/dyn_dnode_msg.c [hunk 5]
```diff
@@ -540,11 +556,19 @@ dyn_parse_req(struct msg *r)
 		if (done_parsing)
 			return;
 
-		if (r->redis) {
+
+		if (r->data_store == DATA_REDIS) {
```
Comment (shailesh33) [inline src/dyn_dnode_msg.c]: wrapper function

File: src/dyn_dnode_msg.c [hunk 6]
```diff
@@ -609,25 +633,41 @@ void dyn_parse_rsp(struct msg *r)
 				mbuf_put(b);
 
 				r->mlen = mbuf_length(decrypted_buf);
-				if (r->redis) {
+
+
+				if (r->data_store==DATA_REDIS) {
```
Comment (shailesh33) [inline src/dyn_dnode_msg.c]: wrapper function

File: src/dyn_dnode_msg.c [hunk 7]
```diff
@@ -609,25 +633,41 @@ void dyn_parse_rsp(struct msg *r)
...
-			//substract alraedy received bytes
+			//Subtract already received bytes
 			dmsg->plen -= b->last - b->pos;
 			return;
 
 		} else if (r->dyn_state == DYN_POST_DONE) {
 			struct mbuf *last_buf = STAILQ_LAST(&r->mhdr, mbuf, next);
 			if (last_buf->read_flip == 1) {
-				if (r->redis)
+				if (r->data_store==DATA_REDIS){
```
Comment (shailesh33) [inline src/dyn_dnode_msg.c]: wrapper function

File: src/dyn_dnode_msg.c [hunk 8]
```diff
@@ -637,11 +677,19 @@ void dyn_parse_rsp(struct msg *r)
 		if (done_parsing)
 			return;
 
-		if (r->redis) {
+		if (r->data_store==DATA_REDIS) {
 			return redis_parse_rsp(r);
```
Comment (shailesh33) [inline src/dyn_dnode_msg.c]: wrapper function

File: src/dyn_dnode_msg.c [hunk 9]
```diff
@@ -637,11 +677,19 @@ void dyn_parse_rsp(struct msg *r)
 		if (done_parsing)
 			return;
 
-		if (r->redis) {
+		if (r->data_store==DATA_REDIS) {
```
Comment (shailesh33) [inline src/dyn_dnode_msg.c]: wrapper function

File: src/dyn_proxy.c [hunk 10]
```diff
@@ -204,9 +204,19 @@ proxy_each_init(void *elem, void *data)
         return status;
     }
 
+    char * log_datastore = "not selected data store";
```
Comment (shailesh33) [inline src/dyn_proxy.c]: if you implement enum, you could have a function to get string based on the type. Less code is good code :p

File: src/dyn_conf.h (hunk 2 above)
Comment (ipapapa) [inline src/dyn_conf.h]: made it an enum

**Comment by shailesh33** [comment]
Hey this is not backword compatible at all. I just cannot drop binary and make it work. What do you think should be done.?

**Pull request merged**